File: runtimes/java/src/main/java/org/findsl/runtime/FinDslAbort.java
// Copyright (C) 2026 devtank42 GmbH
// Licensed under the EUPL-1.2 (see LICENSE) OR a commercial licence
// from devtank42 GmbH (see LICENSE-COMMERCIAL.md).
// SPDX-License-Identifier: EUPL-1.2

package org.findsl.runtime;

/**
 * Spiegel des FinDSL-{@code abbruch}-Ausdrucks (SPEC § 4.19, ADR4) —
 * {@code values.ts AbbruchSignal}.
 *
 * <p>Bewusst eine <b>unchecked</b> {@link RuntimeException}: generierter
 * Code fängt sie nie (analog dazu, dass kein {@code evalExpr}-Pfad sie
 * fängt); sie propagiert bis zur Lauf-Grenze. Generierte JUnit-Tests
 * prüfen einen erwarteten Abbruch über
 * {@code assertThrows(FinDslAbort.class, …)} ({@code erwartet abbruch}).
 */
public final class FinDslAbort extends RuntimeException {

    private static final long serialVersionUID = 1L;

    private final String reason;

    public FinDslAbort(String reason) {
        super("abbruch: " + reason);
        this.reason = reason;
    }

    /** Die Fachbegründung (inkl. evtl. §-Referenz) — ohne Präfix. */
    public String reason() {
        return reason;
    }
}

File: runtimes/ts/src/findsl-abort.ts
// Copyright (C) 2026 devtank42 GmbH
// Licensed under the EUPL-1.2 (see LICENSE) OR a commercial licence
// from devtank42 GmbH (see LICENSE-COMMERCIAL.md).
// SPDX-License-Identifier: EUPL-1.2

/**
 * Spiegel des FinDSL-`abbruch`-Ausdrucks (SPEC § 4.19, ADR4) —
 * `values.ts AbbruchSignal`. 1:1-Port von `org.findsl.runtime.FinDslAbort`.
 *
 * Generierter Code fängt sie nie; sie propagiert bis zur Lauf-Grenze.
 * Generierte Vitest-Tests prüfen einen erwarteten Abbruch via
 * `expect(() => …).toThrow(FinDslAbort)` (`erwartet abbruch`).
 */
export class FinDslAbort extends Error {
    readonly reason: string;

    constructor(reason: string) {
        super('abbruch: ' + reason);
        this.name = 'FinDslAbort';
        this.reason = reason;
    }
}
